File: app/src/main/java/com/njlabs/showjava/processor/JavaExtractor.java
package com.njlabs.showjava.processor;

import com.crashlytics.android.Crashlytics;
import com.njlabs.showjava.utils.SourceInfo;
import com.njlabs.showjava.utils.ZipUtils;
import com.njlabs.showjava.utils.logging.Ln;

import org.benf.cfr.reader.Main;
import org.benf.cfr.reader.state.ClassFileSourceImpl;
import org.benf.cfr.reader.state.DCCommonState;
import org.benf.cfr.reader.util.getopt.GetOptParser;
import org.benf.cfr.reader.util.getopt.Options;
import org.benf.cfr.reader.util.getopt.OptionsImpl;
import org.jetbrains.java.decompiler.main.decompiler.ConsoleDecompiler;
import org.jetbrains.java.decompiler.main.decompiler.PrintStreamLogger;

import java.io.File;
import java.io.PrintStream;
import java.util.HashMap;
import java.util.Map;

import jadx.api.JadxDecompiler;

/**
 * Created by Niranjan on 29-05-2015.
 */
public class JavaExtractor extends ProcessServiceHelper {

    public JavaExtractor(ProcessService processService) {
        this.processService = processService;
        this.UIHandler = processService.UIHandler;
        this.packageFilePath = processService.packageFilePath;
        this.packageName = processService.packageName;
        this.exceptionHandler = processService.exceptionHandler;
        this.sourceOutputDir = processService.sourceOutputDir;
        this.javaSourceOutputDir = processService.javaSourceOutputDir;

        if(printStream == null) {
            printStream = new PrintStream(new ProgressStream());
            System.setErr(printStream);
            System.setOut(printStream);
        }
    }

    public void extract() {

        broadcastStatus("jar2java");

        File dexInputFile = new File(sourceOutputDir + "/optimised_classes.dex");
        File jarInputFile = new File(sourceOutputDir + "/" + packageName + ".jar");

        final File javaOutputDir = new File(javaSourceOutputDir);

        if (!javaOutputDir.isDirectory()) {
            javaOutputDir.mkdirs();
        }

        if(!processService.decompilerToUse.equals("jadx")){
            if(dexInputFile.exists() && dexInputFile.isFile()){
                dexInputFile.delete();
            }
        }

        switch (processService.decompilerToUse) {
            case "jadx":
                decompileWithJaDX(dexInputFile, javaOutputDir);
                break;

            case "cfr":
                decompileWithCFR(jarInputFile, javaOutputDir);
                break;

            case "fernflower":
                decompileWithFernFlower(jarInputFile, javaOutputDir);
                break;

        }

    }

    private void decompileWithCFR(File jarInputFile, File javaOutputDir){
        String[] args = {jarInputFile.toString(), "--outputdir", javaOutputDir.toString()};
        GetOptParser getOptParser = new GetOptParser();

        Options options = null;
        try {
            options = getOptParser.parse(args, OptionsImpl.getFactory());

            if(!options.optionIsSet(OptionsImpl.HELP) && options.getOption(OptionsImpl.FILENAME) != null) {
                ClassFileSourceImpl classFileSource = new ClassFileSourceImpl(options);
                final DCCommonState dcCommonState = new DCCommonState(options, classFileSource);
                final String path = options.getOption(OptionsImpl.FILENAME);

                ThreadGroup group = new ThreadGroup("Jar 2 Java Group");
                Thread javaExtractionThread = new Thread(group, new Runnable() {
                    @Override
                    public void run() {
                        boolean javaError = false;
                        try {
                            Main.doJar(dcCommonState, path);
                        } catch (Exception | StackOverflowError e) {
                            Ln.e(e);
                            javaError = true;
                        }
                        startXMLExtractor(!javaError);
                    }
                }, "Jar to Java Thread", processService.STACK_SIZE);

                javaExtractionThread.setPriority(Thread.MAX_PRIORITY);
                javaExtractionThread.setUncaughtExceptionHandler(exceptionHandler);
                javaExtractionThread.start();

            } else {
                broadcastStatus("exit_process_on_error");
            }

        } catch (Exception e) {
            Crashlytics.logException(e);
            broadcastStatus("exit_process_on_error");
        }

    }

    private void decompileWithJaDX(final File dexInputFile, final File javaOutputDir){

        ThreadGroup group = new ThreadGroup("Jar 2 Java Group");
        Thread javaExtractionThread = new Thread(group, new Runnable() {
            @Override
            public void run() {
                boolean javaError = false;
                try {
                    JadxDecompiler jadx = new JadxDecompiler();
                    jadx.setOutputDir(javaOutputDir);
                    jadx.loadFile(dexInputFile);
                    jadx.saveSources();
                } catch (Exception | StackOverflowError e) {
                    Ln.e(e);
                    javaError = true;
                }
                if(dexInputFile.exists() && dexInputFile.isFile()){
                    dexInputFile.delete();
                }
                startXMLExtractor(!javaError);
            }
        }, "Jar to Java Thread", processService.STACK_SIZE);

        javaExtractionThread.setPriority(Thread.MAX_PRIORITY);
        javaExtractionThread.setUncaughtExceptionHandler(exceptionHandler);
        javaExtractionThread.start();
    }

    private void decompileWithFernFlower(final File jarInputFile, final File javaOutputDir){

        ThreadGroup group = new ThreadGroup("Jar 2 Java Group");
        Thread javaExtractionThread = new Thread(group, new Runnable() {
            @Override
            public void run() {
                boolean javaError = false;
                try {

                    final Map<String, Object> mapOptions = new HashMap<>();
                    PrintStreamLogger logger = new PrintStreamLogger(printStream);
                    ConsoleDecompiler decompiler = new ConsoleDecompiler(javaOutputDir, mapOptions, logger);
                    decompiler.addSpace(jarInputFile, true);
                    decompiler.decompileContext();

                    File decompiledJarFile = new File(javaOutputDir + "/" + packageName +".jar");

                    if(decompiledJarFile.exists()) {
                        ZipUtils.unzip(decompiledJarFile, javaOutputDir, printStream);
                        decompiledJarFile.delete();
                    } else {
                        javaError = true;
                    }

                } catch (Exception | StackOverflowError e) {
                    Ln.e(e);
                    javaError = true;
                }
                startXMLExtractor(!javaError);
            }
        }, "Jar to Java Thread", processService.STACK_SIZE);

        javaExtractionThread.setPriority(Thread.MAX_PRIORITY);
        javaExtractionThread.setUncaughtExceptionHandler(exceptionHandler);
        javaExtractionThread.start();

    }

    private void startXMLExtractor(boolean hasJava) {
        SourceInfo.setjavaSourceStatus(processService, hasJava);
        ((new ResourcesExtractor(processService))).extract();
    }
}

File: app/src/main/kotlin/com/njlabs/showjava/services/processor/handlers/JavaExtractor.kt
package com.njlabs.showjava.services.processor.handlers

import com.njlabs.showjava.services.processor.ProcessorService
import com.njlabs.showjava.utils.PackageSourceTools
import com.njlabs.showjava.utils.ZipUtils
import jadx.api.JadxDecompiler
import org.benf.cfr.reader.Main
import org.benf.cfr.reader.state.ClassFileSourceImpl
import org.benf.cfr.reader.state.DCCommonState
import org.benf.cfr.reader.util.getopt.GetOptParser
import org.benf.cfr.reader.util.getopt.Options
import org.benf.cfr.reader.util.getopt.OptionsImpl
import org.benf.cfr.reader.util.output.DumperFactoryImpl
import org.jetbrains.java.decompiler.main.decompiler.ConsoleDecompiler
import org.jetbrains.java.decompiler.main.decompiler.PrintStreamLogger
import timber.log.Timber
import java.io.File
import java.io.PrintStream


class JavaExtractor(private val processorService: ProcessorService) : BaseHandler(processorService) {
    private var ignoredLibs: ArrayList<String>? = null

    private val sourceOutputDirectory: File = File(processorService.sourceOutputDirectory)
    private val javaSourceOutputDirectory: File = File(processorService.javaSourceOutputDirectory)

    private var dexFile: File = File(sourceOutputDirectory, "optimised_classes.dex")
    private var jarFile: File = File(javaSourceOutputDirectory, "${processorService.inputPackageName}.jar")

    init {
        ignoredLibs = ArrayList()
        printStream = PrintStream(ProgressStream())
        System.setErr(printStream)
        System.setOut(printStream)

        Timber.tag("JavaExtractor")
    }


    fun extract() {
        broadcastStatus("jar2java")

        if (!javaSourceOutputDirectory.isDirectory) {
            javaSourceOutputDirectory.mkdirs()
        }

        if (processorService.decompilerToUse != "jadx") {
            if (dexFile.exists() && dexFile.isFile) {
                dexFile.delete()
            }
        }

        when (processorService.decompilerToUse) {
            "jadx" -> decompileWithJaDX(dexFile, javaSourceOutputDirectory)
            "cfr" -> decompileWithCFR(jarFile, javaSourceOutputDirectory)
            "fernflower" -> decompileWithFernFlower(jarFile, javaSourceOutputDirectory)
        }
    }

    private fun decompileWithCFR(jarInputFile: File, javaOutputDir: File) {
        val args = arrayOf(jarInputFile.toString(), "--outputdir", javaOutputDir.toString())
        val getOptParser = GetOptParser()

        val options: Options?
        try {
            options = getOptParser.parse<Options>(args, OptionsImpl.getFactory())

            if (!options!!.optionIsSet(OptionsImpl.HELP) && options.getOption(OptionsImpl.FILENAME) != null) {
                val classFileSource = ClassFileSourceImpl(options)
                val dcCommonState = DCCommonState(options, classFileSource)
                val path = options.getOption(OptionsImpl.FILENAME)
                val dumperFactory = DumperFactoryImpl()
                var javaError = false
                try {
                    Main.doJar(dcCommonState, path, dumperFactory)
                } catch (e: Exception) {
                    Timber.e(e)
                    javaError = true
                } catch (e: StackOverflowError) {
                    Timber.e(e)
                    javaError = true
                }
                startXMLExtractor(!javaError)

            } else {
                broadcastStatus("exit_process_on_error")
            }

        } catch (e: Exception) {
            Timber.e(e)
            broadcastStatus("exit_process_on_error")
        }

    }

    private fun decompileWithJaDX(dexInputFile: File, javaOutputDir: File) {
        var javaError = false
        try {
            val jadx = JadxDecompiler()
            jadx.setOutputDir(javaOutputDir)
            jadx.loadFile(dexInputFile)
            jadx.saveSources()
        } catch (e: Exception) {
            Timber.e(e)
            javaError = true
        } catch (e: StackOverflowError) {
            Timber.e(e)
            javaError = true
        }

        if (dexInputFile.exists() && dexInputFile.isFile) {
            dexInputFile.delete()
        }
        startXMLExtractor(!javaError)
    }

    private fun decompileWithFernFlower(jarInputFile: File, javaOutputDir: File) {
        var javaError = false
        try {

            val mapOptions = HashMap<String, Any>()
            val logger = PrintStreamLogger(printStream)
            val decompiler = ConsoleDecompiler(javaOutputDir, mapOptions, logger)
            decompiler.addSpace(jarInputFile, true)
            decompiler.decompileContext()

            if (jarFile.exists()) {
                ZipUtils.unzip(jarFile, javaOutputDir, printStream!!)
                jarFile.delete()
            } else {
                javaError = true
            }

        } catch (e: Exception) {
            Timber.e(e)
            javaError = true
        } catch (e: StackOverflowError) {
            Timber.e(e)
            javaError = true
        }

        startXMLExtractor(!javaError)

    }

    private fun startXMLExtractor(hasJava: Boolean) {
        PackageSourceTools.setJavaSourceStatus(sourceOutputDirectory.absolutePath, hasJava)
    }
}